Failed to build latest git version (expected unit struct, unit variant or constant)

Hi!

I read about Biscuit on Twitter, went to the project page (biscuitsec.org) and want to follow the getting-started. For this I want to install `biscuit-cli`. I followed the install instructions on the readme (https://github.com/biscuit-auth/biscuit-cli/tree/3e7e208c8ebe98731e32b4297799101ef789e3b7#installation).
It was not possible for me to build the project this way because of compile errors.
See compile error, git hash, compiler version in the block below:

```
PS C:\Users\stefa\Documents\git> git clone https://github.com/biscuit-auth/biscuit-cli.git
Cloning into 'biscuit-cli'...
remote: Enumerating objects: 141, done.
remote: Counting objects: 100% (141/141), done.
remote: Compressing objects: 100% (77/77), done.
remote: Total 141 (delta 61), reused 107 (delta 34), pack-reused 0 eceiving objects:  86% (122/141)
Receiving objects: 100% (141/141), 39.53 KiB | 2.20 MiB/s, done.
Resolving deltas: 100% (61/61), done.
PS C:\Users\stefa\Documents\git> cd biscuit-cli
PS C:\Users\stefa\Documents\git\biscuit-cli> cargo install --path .
  Installing biscuit-auth-cli v0.1.0 (C:\Users\stefa\Documents\git\biscuit-cli)
    Updating crates.io index
   Compiling version_check v0.9.4
   Compiling proc-macro2 v1.0.36
   Compiling unicode-xid v0.2.2
   Compiling syn v1.0.86
   Compiling autocfg v1.0.1
   Compiling winapi v0.3.9
   Compiling typenum v1.15.0
   Compiling cfg-if v1.0.0
   Compiling getrandom v0.1.16
   Compiling memchr v2.4.1
   Compiling anyhow v1.0.53
   Compiling libc v0.2.113
   Compiling either v1.6.1
   Compiling ppv-lite86 v0.2.16
   Compiling serde v1.0.135
   Compiling subtle v2.4.1
   Compiling signature v1.5.0
   Compiling opaque-debug v0.3.0
   Compiling bytes v1.1.0
   Compiling cpufeatures v0.2.1
   Compiling byteorder v1.4.3
   Compiling heck v0.4.0
   Compiling biscuit-auth v2.0.0
   Compiling regex-syntax v0.6.25
   Compiling hashbrown v0.11.2
   Compiling minimal-lexical v0.2.1
   Compiling hex v0.4.3
   Compiling textwrap v0.14.2
   Compiling lazy_static v1.4.0
   Compiling fastrand v1.7.0
   Compiling strsim v0.10.0
   Compiling bitflags v1.3.2
   Compiling base64 v0.13.0
   Compiling itertools v0.10.3
   Compiling ed25519 v1.3.0
   Compiling generic-array v0.14.5
   Compiling proc-macro-error-attr v1.0.4
   Compiling proc-macro-error v1.0.4
   Compiling nom v7.1.0
   Compiling num-traits v0.2.14
   Compiling num-integer v0.1.44
   Compiling indexmap v1.8.0
   Compiling rand_core v0.5.1
   Compiling aho-corasick v0.7.18
   Compiling os_str_bytes v6.0.0
   Compiling rand_chacha v0.2.2
   Compiling quote v1.0.15
   Compiling rand v0.7.3
   Compiling regex v1.5.4
   Compiling digest v0.9.0
   Compiling block-buffer v0.9.0
   Compiling sha2 v0.9.9
   Compiling time v0.1.44
   Compiling winapi-util v0.1.5
   Compiling atty v0.2.14
   Compiling remove_dir_all v0.5.3
   Compiling chrono v0.4.19
   Compiling termcolor v1.1.2
   Compiling tempfile v3.3.0
   Compiling synstructure v0.12.6
   Compiling zeroize_derive v1.3.1
   Compiling prost-derive v0.9.0
   Compiling thiserror-impl v1.0.30
   Compiling clap_derive v3.0.12
   Compiling zeroize v1.5.0
   Compiling curve25519-dalek v3.2.0
   Compiling thiserror v1.0.30
   Compiling ed25519-dalek v1.0.1
   Compiling prost v0.9.0
   Compiling clap v3.0.12
   Compiling prost-types v0.9.0
   Compiling biscuit-auth-cli v0.1.0 (C:\Users\stefa\Documents\git\biscuit-cli)
error[E0532]: expected unit struct, unit variant or constant, found struct variant `Logic::NoMatchingPolicy`
   --> src/main.rs:652:9
    |
652 |         Logic::NoMatchingPolicy => println!("No allow policy was matched"),
    |         ^^^^^^^^^^^^^^^^^^^^^^^ help: use struct pattern syntax instead: `Logic::NoMatchingPolicy { /* fields */ }`
    |
   ::: C:\Users\stefa\.cargo\registry\src\github.com-1ecc6299db9ec823\biscuit-auth-2.0.0\src\error.rs:177:5
    |
177 |     NoMatchingPolicy {
    |     ---------------- `Logic::NoMatchingPolicy` defined here

error[E0308]: mismatched types
   --> src/main.rs:294:9
    |
294 |       let (remaining, result) = parse_block_source(&string).map_err(|_| E {
    |  _________^^^^^^^^^^^^^^^^^^^___-
    | |         |
    | |         expected struct `SourceResult`, found tuple
295 | |         msg: "parse error".into(),
296 | |     })?;
    | |_______- this expression has type `SourceResult<'_>`
    |
    = note: expected struct `SourceResult<'_>`
                found tuple `(_, _)`

error[E0308]: mismatched types
   --> src/main.rs:331:9
    |
331 |       let (remaining, result) = parse_block_source(&string).map_err(|_| E {
    |  _________^^^^^^^^^^^^^^^^^^^___-
    | |         |
    | |         expected struct `SourceResult`, found tuple
332 | |         msg: "parse error".into(),
333 | |     })?;
    | |_______- this expression has type `SourceResult<'_>`
    |
    = note: expected struct `SourceResult<'_>`
                found tuple `(_, _)`

error[E0308]: mismatched types
   --> src/main.rs:367:9
    |
367 |       let (remaining, result) = parse_source(&string).map_err(|_| E {
    |  _________^^^^^^^^^^^^^^^^^^^___-
    | |         |
    | |         expected struct `SourceResult`, found tuple
368 | |         msg: "parse error".into(),
369 | |     })?;
    | |_______- this expression has type `SourceResult<'_>`
    |
    = note: expected struct `SourceResult<'_>`
                found tuple `(_, _)`

error[E0599]: no variant or associated item named `Deny` found for enum `Logic` in the current scope
   --> src/main.rs:651:16
    |
651 |         Logic::Deny(i) => println!("The deny policy at index {} was matched", i),
    |                ^^^^ variant or associated item not found in `Logic`

error[E0599]: no variant or associated item named `FailedChecks` found for enum `Logic` in the current scope
   --> src/main.rs:653:16
    |
653 |         Logic::FailedChecks(checks) => {
    |                ^^^^^^^^^^^^ variant or associated item not found in `Logic`

Some errors have detailed explanations: E0308, E0532, E0599.
For more information about an error, try `rustc --explain E0308`.
error: failed to compile `biscuit-auth-cli v0.1.0 (C:\Users\stefa\Documents\git\biscuit-cli)`, intermediate artifacts can be found at `C:\Users\stefa\Documents\git\biscuit-cli\target`

Caused by:
  could not compile `biscuit-auth-cli` due to 6 previous errors
PS C:\Users\stefa\Documents\git\biscuit-cli> git rev-parse HEAD
3e7e208c8ebe98731e32b4297799101ef789e3b7
PS C:\Users\stefa\Documents\git\biscuit-cli> cargo version -v
cargo 1.58.0 (f01b232bc 2022-01-19)
release: 1.58.0
commit-hash: f01b232bc7f4d94f0c4603930a5a96277715eb8c
commit-date: 2022-01-19
host: x86_64-pc-windows-msvc
libgit2: 1.3.0 (sys:0.13.23 vendored)
libcurl: 7.80.0-DEV (sys:0.4.51+curl-7.80.0 vendored ssl:Schannel)
os: Windows 10.0.22000 (Windows 10 Pro) [64-bit]
```

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -15,7 +15,7 @@ path = "src/main.rs"
 
 [dependencies]
 atty = "0.2.14"
-biscuit-auth = "2.0.0-beta5"
+biscuit-auth = "2.0.0"
 clap = { version = "3.0.0-rc.7", features = ["color", "derive"] }
 hex = "0.4.3"
 tempfile = "3.2.0"

--- a/src/main.rs
+++ b/src/main.rs
@@ -1,7 +1,7 @@
 use atty::Stream;
 use biscuit_auth::{
-    builder::{BiscuitBuilder, BlockBuilder},
-    error::{FailedCheck, Logic, RunLimit, Token},
+    builder::{BiscuitBuilder, BlockBuilder, Policy},
+    error::{FailedCheck, Logic, MatchedPolicy, RunLimit, Token},
     parser::{parse_block_source, parse_source},
     Authorizer, Biscuit, UnverifiedBiscuit, {KeyPair, PrivateKey, PublicKey},
 };
@@ -291,15 +291,10 @@ fn read_authority_from(
         DatalogInput::FromFile(f) => fs::read_to_string(&f)?,
         DatalogInput::DatalogString(str) => str.to_owned(),
     };
-    let (remaining, result) = parse_block_source(&string).map_err(|_| E {
+    let result = parse_block_source(&string).map_err(|_| E {
         msg: "parse error".into(),
     })?;
-    if !remaining.is_empty() {
-        return Err(E {
-            msg: "remaining input".into(),
-        }
-        .into());
-    }
+
     for (fact_str, _) in result.facts {
         builder.add_authority_fact(fact_str)?;
     }
@@ -328,15 +323,10 @@ fn read_block_from(
         DatalogInput::FromFile(f) => fs::read_to_string(&f)?,
         DatalogInput::DatalogString(str) => str.to_owned(),
     };
-    let (remaining, result) = parse_block_source(&string).map_err(|_| E {
+    let result = parse_block_source(&string).map_err(|_| E {
         msg: "parse error".into(),
     })?;
-    if !remaining.is_empty() {
-        return Err(E {
-            msg: "remaining input".into(),
-        }
-        .into());
-    }
+
     for (fact_str, _) in result.facts {
         builder.add_fact(fact_str)?;
     }
@@ -364,15 +354,10 @@ fn read_authorizer_from(
         DatalogInput::FromFile(f) => fs::read_to_string(&f)?,
         DatalogInput::DatalogString(str) => str.to_owned(),
     };
-    let (remaining, result) = parse_source(&string).map_err(|_| E {
+    let result = parse_source(&string).map_err(|_| E {
         msg: "parse error".into(),
     })?;
-    if !remaining.is_empty() {
-        return Err(E {
-            msg: "remaining input".into(),
-        }
-        .into());
-    }
+
     for (fact_str, _) in result.facts {
         authorizer.add_fact(fact_str)?;
     }
@@ -617,13 +602,21 @@ fn handle_inspect(inspect: &Inspect) -> Result<(), Box<dyn Error>> {
         if let Some(auth_from) = authorizer_from {
             let mut authorizer_builder = biscuit.authorizer()?;
             read_authorizer_from(&auth_from, &mut authorizer_builder)?;
+            let (_, _, _, policies) = authorizer_builder.dump();
             let authorizer_result = authorizer_builder.authorize();
             match authorizer_result {
-                Ok(_) => println!("✅ Authorizer check succeeded 🛡️"),
+                Ok(i) => {
+                    println!("✅ Authorizer check succeeded 🛡️");
+                    println!(
+                        "Matched allow policy: {}",
+                        policies.get(i).expect("Incorrect policy index")
+                    );
+                }
+
                 Err(e) => {
                     println!("❌ Authorizer check failed 🛡️");
                     match e {
-                        Token::FailedLogic(l) => display_logic_error(&l),
+                        Token::FailedLogic(l) => display_logic_error(&policies, &l),
                         Token::RunLimit(l) => display_run_limit(&l),
                         _ => {}
                     }
@@ -646,19 +639,43 @@ fn handle_inspect(inspect: &Inspect) -> Result<(), Box<dyn Error>> {
     Ok(())
 }
 
-fn display_logic_error(e: &Logic) {
+fn display_logic_error(policies: &[Policy], e: &Logic) {
     match e {
-        Logic::Deny(i) => println!("The deny policy at index {} was matched", i),
-        Logic::NoMatchingPolicy => println!("No allow policy was matched"),
-        Logic::FailedChecks(checks) => {
-            println!("The following checks failed:");
-            display_failed_checks(checks);
+        Logic::Unauthorized { policy, checks } => {
+            display_matched_policy(policies, &policy);
+            display_failed_checks(&checks);
+        }
+        Logic::NoMatchingPolicy { checks } => {
+            println!("No policy matched");
+            display_failed_checks(&checks);
         }
         e => println!("An execution error happened during authorization: {:?}", &e),
     }
 }
 
+fn display_matched_policy(policies: &[Policy], policy: &MatchedPolicy) {
+    match policy {
+        MatchedPolicy::Allow(i) => {
+            let policy = policies.get(*i);
+            println!(
+                "An allow policy matched: {}",
+                policy.expect("Incorrect policy index")
+            );
+        }
+        MatchedPolicy::Deny(i) => {
+            let policy = policies.get(*i);
+            println!(
+                "A deny policy matched: {}",
+                policy.expect("Incorrect policy index")
+            );
+        }
+    }
+}
+
 fn display_failed_checks(checks: &Vec<FailedCheck>) {
+    if checks.len() > 0 {
+        println!("The following checks failed:");
+    }
     for c in checks {
         match c {
             FailedCheck::Block(bc) => {
@@ -667,9 +684,9 @@ fn display_failed_checks(checks: &Vec<FailedCheck>) {
                 } else {
                     format!("Block {}", &bc.block_id)
                 };
-                println!("{} check: {}", &block_name, &bc.rule);
+                println!("  {} check: {}", &block_name, &bc.rule);
             }
-            FailedCheck::Authorizer(ac) => println!("Authorizer check: {}", &ac.rule),
+            FailedCheck::Authorizer(ac) => println!("  Authorizer check: {}", &ac.rule),
         }
     }
 }

